### PR TITLE
SSL/TLSize all the things! (convert http:// to https:// where possible) (take two)

### PR DESCRIPTION
**Pull request by reedloden**
Even though ruby-build uses hardcoded checksums to help prevent MITM and corruption issues, using SSL/TLS to download the files adds another layer of protection, ensuring that nobody can see that somebody is downloading a certain ruby version.

This is a follow-up to #750 to address some of the comments that @mislav had with it and also to deal with Ruby proper just recently enabling SSL for their downloads (https://bugs.ruby-lang.org/issues/10672#change-53270). Also, I checked with the rubygems.org folk, and they are deprecating CloudFront and wanting people to use Fastly for download URLs (see https://github.com/rubygems/rubygems-infrastructure/issues/25).

### CODE REVIEW

**Comment by reedloden** [comment]
Getting a test failure since it's trying to resolve files.rubyforge.vm.bytemark.co.uk, which doesn't exist anymore... Let me know what you want to do there (could remove those entirely since they don't work).

File: test/build.bats [hunk 1]
```diff
@@ -510,7 +510,7 @@ OUT
   cached_tarball "jruby-1.7.9" bin/foo.exe bin/bar.dll bin/baz.bat
 
   run_inline_definition <<DEF
-install_package "jruby-1.7.9" "http://jruby.org/downloads/jruby-bin-1.7.9.tar.gz" jruby
+install_package "jruby-1.7.9" "https://s3.amazonaws.com/jruby-org/downloads/1.7.9/jruby-bin-1.7.9.tar.gz" jruby
```
Comment (reedloden) [inline test/build.bats]: This was changed, as this is the URL that `share/ruby-build/jruby-1.7.9` uses.

**Comment by mislav** [comment]
Thanks for revisiting this after informing us that Ruby has enabled HTTPS on its own official download locations. I will leave my comments inline, but in general, I feel like this changeset needs to be split in several different commits, and that changes unrelated to HTTPS protocol switch (like correcting typos) don't belong to this PR.

File: README.md [hunk 2]
```diff
@@ -147,7 +147,7 @@ Example usage:
 $ rbenv install --patch 1.9.3-p429 < /path/to/ruby.patch
 
 # applying a patch from HTTP
-$ rbenv install --patch 1.9.3-p429 < <(curl -sSL http://git.io/ruby.patch)
+$ rbenv install --patch 1.9.3-p429 < <(curl -sSL https://git.io/ruby.patch)
```
Comment (mislav) [inline README.md]: This change wasn't necessary. The URL is for example only, and doesn't even resolve to anything (`ruby.patch` is a 404).

File: README.md [hunk 3]
```diff
@@ -179,7 +179,7 @@ mirror and use official URLs instead. You can force ruby-build to bypass the
 mirror by setting the `RUBY_BUILD_SKIP_MIRROR` environment variable.
 
 The official ruby-build download mirror is sponsored by
-[37signals](http://37signals.com/).
+[Basecamp](https://basecamp.com/).
```
Comment (mislav) [inline README.md]: This correction doesn't belong to this PR.

File: share/ruby-build/1.8.6-p383 [hunk 4]
```diff
@@ -1,3 +1,3 @@
 require_gcc
-install_package "ruby-1.8.6-p383" "http://cache.ruby-lang.org/pub/ruby/1.8/ruby-1.8.6-p383.tar.gz#cea6c67f737727007ec89d1c93fd7d0dba035220f981706091b8642d7a43c03a" auto_tcltk standard
-install_package "rubygems-1.3.7" "http://production.cf.rubygems.org/rubygems/rubygems-1.3.7.tgz#388b90ae6273f655507b10c8ba6bee9ea72e7d49c3e610025531cb8c3ba67c9d" ruby
+install_package "ruby-1.8.6-p383" "https://cache.ruby-lang.org/pub/ruby/1.8/ruby-1.8.6-p383.tar.gz#cea6c67f737727007ec89d1c93fd7d0dba035220f981706091b8642d7a43c03a" auto_tcltk standard
+install_package "rubygems-1.3.7" "https://rubygems.global.ssl.fastly.net/rubygems/rubygems-1.3.7.tgz#388b90ae6273f655507b10c8ba6bee9ea72e7d49c3e610025531cb8c3ba67c9d" ruby
```
Comment (mislav) [inline share/ruby-build/1.8.6-p383]: Like @sferik, I'm reluctant to switch official download locations away from `rubygems.org` domain and onto `fastly.net`. Even if the Rubygems team switches to Fastly going forward, are they going to at least provide canonical download resources under `rubygems.org` that result in a redirect to a CDN like Fastly?

In any case, this switch should be done as a separate commit because it's a separate topic that we might or might not accept at this point.

File: share/ruby-build/ree-1.8.7-2011.03 [hunk 5]
```diff
@@ -1,3 +1,3 @@
 require_gcc
-install_package "ruby-enterprise-1.8.7-2011.03" "http://rubyenterpriseedition.googlecode.com/files/ruby-enterprise-1.8.7-2011.03.tar.gz#0c0ddbc43b3aef49686db27e761e55a23437f12e1f00b6fe55d94724637bff6b" ree_installer
-install_package "rubygems-1.6.2" "http://production.cf.rubygems.org/rubygems/rubygems-1.6.2.tgz#cb5261818b931b5ea2cb54bc1d583c47823543fcf9682f0d6298849091c1cea7" ruby
+install_package "ruby-enterprise-1.8.7-2011.03" "https://rubyenterpriseedition.googlecode.com/files/ruby-enterprise-1.8.7-2011.03.tar.gz#0c0ddbc43b3aef49686db27e761e55a23437f12e1f00b6fe55d94724637bff6b" ree_installer
```
Comment (mislav) [inline share/ruby-build/ree-1.8.7-2011.03]: I would also like to see REE-related changes be done as a separate commit as well.

File: test/arguments.bats [hunk 6]
```diff
@@ -2,7 +2,7 @@
 
 load test_helper
 
-@test "not enought arguments for ruby-build" {
+@test "not enough arguments for ruby-build" {
```
Comment (mislav) [inline test/arguments.bats]: Nice catch, but doesn't belong to this PR. Let's try to keep the diffs small.

File: test/build.bats [hunk 7]
```diff
@@ -164,7 +164,7 @@ OUT
   stub_make_install
 
   run_inline_definition <<DEF
-install_package "ruby-2.0.0" "http://ruby-lang.org/ruby/2.0/ruby-2.0.0.tar.gz"
+install_package "ruby-2.0.0" "https://cache.ruby-lang.org/pub/ruby/2.0/ruby-2.0.0.tar.gz"
```
Comment (mislav) [inline test/build.bats]: This change to tests wasn't necessary. The test can continue to keep using HTTP resources since they are fake anway. It's not like the tests are "insecure".

File: test/build.bats (hunk 1 above)
Comment (mislav) [inline test/build.bats]: Again, I'm sad that the canonical download location for JRuby isn't under the `jruby.org` domain anymore. But I guess there's nothing we can do since they seem to be using `s3.amazonaws.com` on their official download page anyway.

**Comment by mislav** [comment]
Feel free to update this PR by splitting this large commit into a series of separate commits that each handle one topic (one for Ruby download locations, another for Rubygems, another for REE, another for JRuby…) and force-pushing to this same branch. Also, please leave out the stuff that I didn't feel belongs to this PR. Thanks for doing the work! :bow:

**Comment by mislav** [comment]
I've removed old REE versions that aren't downloadable anymore. Please rebase to master to get rid of CI failures.

**Comment by reedloden** [comment]
As requested, I've split this up into 5 separate commits. I'll file separate PRs for the README and typo fixes.

File: share/ruby-build/1.8.6-p383 (hunk 4 above)
Comment (reedloden) [inline share/ruby-build/1.8.6-p383]: I will see if they can set something up (even just a redirect at their nginx config), but this needs to change anyway, as the rubygems.org folk have said that they are deprecating CloudFront (which is production.cf.rubygems.org) in favor of Fastly.

**Comment by mislav** [comment]
I hope this resolves this year-old issue. Thanks to:

* @boourns for [initially suggesting it](https://github.com/sstephenson/ruby-build/pull/611).
* @paxan, whose original commit 3cee1b9 I've included here because [he argued for the feature](https://github.com/sstephenson/ruby-build/pull/623).
* @konklone for pushing the Ruby core team to enable HTTPS on their domain.
* And of course @reedloden for doing the hard work of coding all this up through several iterations.

:+1: :bow:

**Comment by konklone** [comment]
And huge thanks to you, @mislav, for incorporating our work and for pushing us to go push the Ruby project upstream -- which will improve things for more than just `rbenv` users.

:+1: :confetti_ball: :lock:

**Comment by reedloden** [comment]
woot :)

**Comment by mislav** [comment]
@reedloden Please follow-up if you hear of any updates re: Rubygems and JRuby canonical download location. We now link to CDNs directly but it would be great if in the long run the download locations were somehow under rubygems.org and jruby.org.

**Comment by reedloden** [comment]
@mislav Will do. Also talking to rubinius guys about their download URLs as well.

**Comment by paxan** [comment]
:+1:
